app/src/main/java/com/vci/vectorcamapp/core/data/upload/image/ImageUploadWorker.kt

```diff
@@ -2,85 +2,508 @@ package com.vci.vectorcamapp.core.data.upload.image
+                    e
+                )
+                return when (val existingUrlResult =
+                    imageUploadDataSource.imageExists(uploadSpecimenId, md5)) {
+                    is DomainResult.Success -> DomainResult.Success(existingUrlResult.data.toString())
+                    is DomainResult.Error -> {
+                        Log.e(
+                            "ImageUploadWorker",
+                            "MD5 check failed despite 409. Upload initialization failed.",
+                            e
+                        )
+                        DomainResult.Error(existingUrlResult.error)
+                    }
+                }
+            } else if (e.shouldRetry()) {
+                Log.w(
+                    "ImageUploadWorker",
+                    "resumeOrCreateUpload failed with a retryable server error.",
+                    e
+                )
+                return DomainResult.Error(NetworkError.SERVER_ERROR)
+            } else {
+                Log.e(
+                    "ImageUploadWorker",
+                    "resumeOrCreateUpload failed with a non-retryable client error.",
+                    e
+                )
+                return DomainResult.Error(NetworkError.CLIENT_ERROR)
+            }
+        } catch (e: IOException) {
+            Log.e("ImageUploadWorker", "resumeOrCreateUpload failed with IOException", e)
+            return DomainResult.Error(NetworkError.NO_INTERNET)
+        }
+
+        val uploader =
+            uploaderResult.successOrNull() ?: return DomainResult.Error(NetworkError.UNKNOWN_ERROR)
+
+        val loopResult = executeUploadLoop(uploader, tusClient, upload)
+        if (loopResult is DomainResult.Error) {
+            return DomainResult.Error(loopResult.error)
+        }
+
+        return when (val finalUrlResult = safeFinish(uploader)) {
+            is DomainResult.Success -> {
+                updateProgressNotification("Verifying...")
+                Log.d("ImageUploadWorker", "Upload finished successfully: ${finalUrlResult.data}")
+                DomainResult.Success(finalUrlResult.data.toString())
+            }
+
+            is DomainResult.Error -> DomainResult.Error(finalUrlResult.error)
+        }
+    }
+
+    private suspend fun executeUploadLoop(
+        initialUploader: TusUploader,
+        client: TusClient,
+        upload: TusUpload
+    ): DomainResult<Unit, NetworkError> {
+        var uploader = initialUploader
+        var currentChunkSize = INITIAL_CHUNK_SIZE_BYTES
+        var successfulUploadsInARow = 0
+        var recoveryAttempts = 0
+
+        while (uploader.offset < upload.size) {
+            uploader.chunkSize = currentChunkSize
+            uploader.requestPayloadSize = currentChunkSize
+
+            val chunkResult = try {
+                withTimeout(NETWORK_TIMEOUT_MS) { DomainResult.Success(uploader.uploadChunk()) }
+            } catch (e: Exception) {
+                when (e) {
```

Comment:
Apparently, instead of having a when statement here, you should just have mutliple catch statements. Check out how this is done in safeCall.